Project 1: DNA

tikhonbelousko opened the pull request.

Commit:
Project 1: DNA

Commit:
Improved mutations

Now mutations work with cells of any size.

Commit:
Added assertion that percentage value is in range from 0 to 100

Commit:
Cell size changed from 20 to 100

Commit:
FINAL

Comments added.

iredchuk commented:
Переменную symbols нужно объявить в @implementation - она не должна быть видна пользователям класса.
description не нужно объявлять в @interface класса Cell - это стандартный метод NSObject. То же с init.

[NSMutableArray arrayWithObjects:@"A", @"T", @"G", @"C", nil]; - это массив объектов NSString, для заполнения именно символами должно было бы быть @'A', @'T', @'G', @'C'. Разница в данном случае не существенна, но вообще говоря char оптимальнее в смысле перфоманса.

Категории лучше создавать в отдельных файлах - так код будет читаться лучше.

Метод randomObject в категории NSMutableArray (Random) крешнется с необработанным исключением, если массив пустой.

В методе mutate массив indexes можно заполнить номерами от 0 до countOfMutations и в цикле просто написать         long indexToChage = [[indexes objectAtIndex:i] integerValue]; - вместо lastObject и последующего удаления объектов из массива

В методе mutate массив symbols копируется на каждой итерации цикла, что не очень оптимально, если предположить, что массив допустимых символов может стать достаточно большим.

tikhonbelousko commented:
Ага, всё поправлю. Спасибо. Только вот как избежать копирования? Подбирать символ в цикле пока он не станет отличным от текущего?

iredchuk commented:
Как вариант, можно взять просто следующий элемент массива, но это не совсем честно, поскольку вероятность мутировать на следующий элемент будет выше, чем на остальные (хотя я так и сделал :))
А можно взять индекс изменяемого символа в массиве symbols, прибавить к нему случайное число от 1 до 3 и у результата взять % 4  - это будет индекс элемента, на который заменяем. Сорри, что пишу словами - отвечаю с iPada.

tikhonbelousko commented:
Ну со следующим элементом было совсем нечестно. Исправил на второй вариант. :) 

Проблема с переменной **symbols**. Когда перенёс её в имплементацию, она стала не видна в категории **Mutator**. Не понимаю почему так. Исправил на **readonly** свойство, как сделать нормально не знаю, но так хоть компилируется.

То что **randomObject** крешится на пустом массиве, я думаю это нормальное поведение. Просто добавил больше информации в эксепшн.

iredchuk commented:
Категория не видит приватных членов класса, это нормально. У категорий доступ к основному интерфейсу класса такой же, как и у подклассов. К сожалению, в Objective C нет <code>protected</code> свойств и методов (которые были бы видны только в подклассах и категориях, но не из других классов). Так что readonly-свойство - в принципе приемлемый вариант, хоть и не очень красивый. Так же как и <code>DNA</code>, которое тоже не должно быть по идее видно как public.

Еще вариант, как я сделал - создал отдельный класс <code>DNAGenerator</code> и в нем инкапсулировал метод генерации случайных символов, так что в нем массив <code>symbols</code> приватный. При этом класс <code>Cell</code> и категория <code>Cell (Mutator)</code>  вызывают методы <code>DNAGenerator</code>-а.

Commit:
FINAL 2.0

- Mutator category in separate file.
- Now symbols variable is a read-only property.
- Some code improvements in Mutator category.
